src/nlp/arrow_dataset.py

```diff
@@ -606,15 +691,22 @@ def apply_function_on_filtered_inputs(inputs, indices):
 
         # Loop over single examples or batches and write to buffer/file if examples are to be updated
         if not batched:
-            for i, example in tqdm(enumerate(self)):
+            for i, example in enumerate(tqdm(self)):
                 example = apply_function_on_filtered_inputs(example, i)
                 if update_data:
                     writer.write(example)
         else:
             for i in tqdm(range(0, len(self), batch_size)):
                 batch = self[i : i + batch_size]
                 indices = list(range(*(slice(i, i + batch_size).indices(self._data.num_rows))))  # Something simpler?
-                batch = apply_function_on_filtered_inputs(batch, indices)
+                try:
+                    batch = apply_function_on_filtered_inputs(
+                        batch, indices, check_same_num_examples=len(self.list_indexes()) > 0
+                    )
+                except NumExamplesMismatch:
+                    raise DatasetTransformationNotAllowedError(
+                        "Using `.map` in batched mode on a dataset with attached indexes is allowed only if it doesn't create or remove existing examples."
+                    )
```

Comment:
Clean!

Maybe you can add that the user can `drop_index` before doing so and recreate the index afterward (and same for all the other places where you forbid operating on a dataset with an index)